### PR TITLE
Automatically detect when the ES operator is available

### LINKED ISSUE #238: Enable provisioning of Elasticsearch only when the ES operator is available
Currently, the Jaeger Operator attempts to provision an Elasticsearch cluster even when the Elasticsearch Operator isn't available. It should work similar to the Routes, where it's enabled/disabled via a flag, with the default being the automatic detection.

### PR DESCRIPTION
**Pull request by jpkrohling**
Closes #238.

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

**Comment by jpkrohling** [comment]
This change is [Reviewable](https://reviewable.io/reviews/jaegertracing/jaeger-operator/239)

File: pkg/apis/io/v1alpha1/jaeger_types.go [hunk 1]
```diff
@@ -21,6 +21,15 @@ const (
 	// FlagPlatformAutoDetect represents the "auto-detect" value for the platform flag
 	FlagPlatformAutoDetect = "auto-detect"
 
+	// FlagProvisionElasticsearchAuto represents the 'auto' value for the 'es-provision' flag
+	FlagProvisionElasticsearchAuto = "auto"
```
Comment (pavolloffay) [inline pkg/apis/io/v1alpha1/jaeger_types.go]: more idiomatic is making it a type

File: pkg/cmd/start/main.go [hunk 2]
```diff
@@ -102,22 +106,39 @@ func start(cmd *cobra.Command, args []string) {
...
+
+			log.WithField("platform", viper.GetString("platform")).Info("Auto-detected the platform")
+		} else {
+			log.WithField("platform", viper.GetString("platform")).Debug("The 'platform' option is explicitly set")
 		}
 
-		if os {
-			viper.Set("platform", v1alpha1.FlagPlatformOpenShift)
+		// detect whether the Elasticsearch operator is available
+		if strings.EqualFold(viper.GetString("es-provision"), v1alpha1.FlagProvisionElasticsearchAuto) {
```
Comment (pavolloffay) [inline pkg/cmd/start/main.go]: This requires deploying ES operator before Jaeger operator which creates assumptions on order how operators are deployed by OLM and get to init state. Do you know if OLM is capable of resolving this?

File: pkg/controller/jaeger/jaeger_controller.go [hunk 3]
```diff
@@ -132,8 +132,10 @@ func defaultStrategyChooser(instance *v1alpha1.Jaeger) strategy.S {
 }
 
 func (r *ReconcileJaeger) apply(jaeger v1alpha1.Jaeger, str strategy.S) error {
-	if err := r.applyElasticsearches(jaeger, str.Elasticsearches()); err != nil {
-		return err
+	if strings.EqualFold(viper.GetString("es-provision"), v1alpha1.FlagProvisionElasticsearchTrue) {
```
Comment (pavolloffay) [inline pkg/controller/jaeger/jaeger_controller.go]: I would rather always check if the es operator is running and if it's not I would log an error here. 

* It's a simpler change
* it explicitly tells user something is wrong, whereas with the current approach logs a message at init time

File: pkg/controller/jaeger/jaeger_controller.go [hunk 4]
```diff
@@ -132,8 +132,10 @@ func defaultStrategyChooser(instance *v1alpha1.Jaeger) strategy.S {
...
+		if err := r.applyElasticsearches(jaeger, str.Elasticsearches()); err != nil {
```
Comment (pavolloffay) [inline pkg/controller/jaeger/jaeger_controller.go]: There should be a log message if strategy contains ES objects and ES API is not available.

File: pkg/controller/jaeger/jaeger_controller.go (hunk 3 above)
Comment (jpkrohling) [reply pkg/controller/jaeger/jaeger_controller.go]: > check if the es operator is running

That's not really an easy thing to do. We can't know for sure *where* the operator is running. We should really just be checking if the cluster understands what a "Elasticsearch" CR is.

File: pkg/cmd/start/main.go (hunk 2 above)
Comment (jpkrohling) [reply pkg/cmd/start/main.go]: Technically, only the CRD has to be deployed, but you have a good point. In cases where people know that there is/will be an operator running, they can set `es-provision` to true.

File: pkg/apis/io/v1alpha1/jaeger_types.go (hunk 1 above)
Comment (jpkrohling) [reply pkg/apis/io/v1alpha1/jaeger_types.go]: Our flags in general need some love as well. For this PR, I'll follow what we already have there for consistency, but will open an issue to change the code base to use types wherever it makes sense.

Comment (jpkrohling) [reply pkg/apis/io/v1alpha1/jaeger_types.go]: In fact, I think we have one such item already: https://github.com/jaegertracing/jaeger-operator/issues/59

File: pkg/controller/jaeger/jaeger_controller.go (hunk 3 above)
Comment (pavolloffay) [reply pkg/controller/jaeger/jaeger_controller.go]: I mean cluster understand's ES CR...

File: pkg/apis/io/v1alpha1/jaeger_types.go (hunk 1 above)
Comment (pavolloffay) [reply pkg/apis/io/v1alpha1/jaeger_types.go]: The referenced issue talks about constants which these already are.

File: pkg/controller/jaeger/jaeger_controller.go [hunk 5]
```diff
@@ -132,8 +132,16 @@ func defaultStrategyChooser(instance *v1alpha1.Jaeger) strategy.S {
...
+	elasticsearches := str.Elasticsearches()
+	if strings.EqualFold(viper.GetString("es-provision"), v1alpha1.FlagProvisionElasticsearchTrue) {
+		if err := r.applyElasticsearches(jaeger, elasticsearches); err != nil {
+			return err
+		}
+	} else if len(elasticsearches) > 0 {
+		log.WithFields(log.Fields{
+			"namespace": jaeger.Namespace,
+			"instance":  jaeger.Name,
+		}).Info("An Elasticsearch cluster should be provisioned, but provisioning is disabled for this Jaeger Operator")
```
Comment (pavolloffay) [inline pkg/controller/jaeger/jaeger_controller.go]: Might be better to use warn level

File: pkg/apis/io/v1alpha1/jaeger_types.go (hunk 1 above)
Comment (jpkrohling) [reply pkg/apis/io/v1alpha1/jaeger_types.go]: I changed #59 to include the "type" part.

**Comment by jpkrohling** [comment]
Tests are now passing again:

```
$ make test
Running unit tests...
?   	github.com/jaegertracing/jaeger-operator/cmd	[no test files]
?   	github.com/jaegertracing/jaeger-operator/cmd/manager	[no test files]
ok  	github.com/jaegertracing/jaeger-operator/pkg/account	0.049s	coverage: 100.0% of statements
?   	github.com/jaegertracing/jaeger-operator/pkg/apis	[no test files]
ok  	github.com/jaegertracing/jaeger-operator/pkg/apis/io/v1alpha1	0.022s	coverage: 15.7% of statements
?   	github.com/jaegertracing/jaeger-operator/pkg/cmd/start	[no test files]
?   	github.com/jaegertracing/jaeger-operator/pkg/cmd/version	[no test files]
ok  	github.com/jaegertracing/jaeger-operator/pkg/config/sampling	0.015s	coverage: 94.1% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/config/ui	0.048s	coverage: 94.1% of statements
?   	github.com/jaegertracing/jaeger-operator/pkg/controller	[no test files]
?   	github.com/jaegertracing/jaeger-operator/pkg/controller/deployment	[no test files]
ok  	github.com/jaegertracing/jaeger-operator/pkg/controller/jaeger	2.190s	coverage: 73.1% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/cronjob	0.050s	coverage: 94.4% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/deployment	0.027s	coverage: 100.0% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/ingress	0.041s	coverage: 100.0% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/inject	0.013s	coverage: 100.0% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/inventory	0.052s	coverage: 91.8% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/route	0.025s	coverage: 100.0% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/service	0.024s	coverage: 100.0% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/storage	0.510s	coverage: 94.5% of statements
?   	github.com/jaegertracing/jaeger-operator/pkg/storage/elasticsearch/v1alpha1	[no test files]
ok  	github.com/jaegertracing/jaeger-operator/pkg/strategy	0.016s	coverage: 91.8% of statements
ok  	github.com/jaegertracing/jaeger-operator/pkg/util	0.006s	coverage: 100.0% of statements
?   	github.com/jaegertracing/jaeger-operator/pkg/version	[no test files]
Formatting code...
Building...
Sending build context to Docker daemon  131.5MB
Step 1/6 : FROM centos
 ---> 1e1148e4cc2c
Step 2/6 : RUN INSTALL_PKGS="       openssl       " &&     yum install -y $INSTALL_PKGS &&     rpm -V $INSTALL_PKGS &&     yum clean all &&     mkdir /tmp/_working_dir &&     chmod og+w /tmp/_working_dir
 ---> Using cache
 ---> 82728e7bfc7b
Step 3/6 : COPY scripts/* /scripts/
 ---> Using cache
 ---> 42400106dc1c
Step 4/6 : USER nobody
 ---> Using cache
 ---> a590f9fcf745
Step 5/6 : ADD build/_output/bin/jaeger-operator /usr/local/bin/jaeger-operator
 ---> ddd6a9aee680
Step 6/6 : ENTRYPOINT ["/usr/local/bin/jaeger-operator"]
 ---> Running in a94737220a06
Removing intermediate container a94737220a06
 ---> 79fae476252a
Successfully built 79fae476252a
Successfully tagged jpkroehling/jaeger-operator:latest
Pushing image jpkroehling/jaeger-operator:latest...
Running end-to-end tests...
ok  	github.com/jaegertracing/jaeger-operator/test/e2e	292.453s
```

**Pull request merged**